File: scanner-go/internal/services/urlscan.go
package services

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/Ahlyx/scanner-go/internal/models"
)

const (
	urlscanSubmitURL = "https://urlscan.io/api/v1/scan/"
	urlscanResultURL = "https://urlscan.io/api/v1/result/"
)

func FetchURLScan(apiKey, targetURL string) (*models.URLScanData, models.SourceMetadata) {
	meta := models.SourceMetadata{Source: "urlscan", RetrievedAt: time.Now().UTC()}

	acquireSem()
	defer releaseSem()

	// Submit scan
	payload, _ := json.Marshal(map[string]string{
		"url":        targetURL,
		"visibility": "public",
	})

	submitReq, err := http.NewRequest(http.MethodPost, urlscanSubmitURL, bytes.NewReader(payload))
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	submitReq.Header.Set("API-Key", apiKey)
	submitReq.Header.Set("Content-Type", "application/json")

	submitResp, err := httpClient.Do(submitReq)
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	defer submitResp.Body.Close()

	if submitResp.StatusCode != http.StatusOK && submitResp.StatusCode != 400 {
		// 400 can indicate already-queued scan; treat non-200/400 as errors.
		if submitResp.StatusCode >= 400 {
			meta.Error = ptr(fmt.Sprintf("HTTP %d", submitResp.StatusCode))
			return nil, meta
		}
	}

	submitBody, err := io.ReadAll(submitResp.Body)
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	var submitData struct {
		UUID string `json:"uuid"`
	}
	if err := json.Unmarshal(submitBody, &submitData); err != nil || submitData.UUID == "" {
		errMsg := "no scan UUID returned"
		if err != nil {
			errMsg = err.Error()
		}
		meta.Error = ptr(errMsg)
		return nil, meta
	}

	// Wait for scan to complete — mirrors Python's asyncio.sleep(15).
	time.Sleep(15 * time.Second)

	// Fetch result
	resultReq, err := http.NewRequest(http.MethodGet,
		fmt.Sprintf("%s%s/", urlscanResultURL, submitData.UUID), nil)
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	resultResp, err := httpClient.Do(resultReq)
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	defer resultResp.Body.Close()

	if resultResp.StatusCode != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", resultResp.StatusCode))
		return nil, meta
	}

	resultBody, err := io.ReadAll(resultResp.Body)
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	var raw struct {
		Verdicts struct {
			Overall struct {
				Malicious bool `json:"malicious"`
				Score     int  `json:"score"`
			} `json:"overall"`
			URLScan struct {
				Categories []string `json:"categories"`
			} `json:"urlscan"`
		} `json:"verdicts"`
		Task struct {
			ScreenshotURL *string `json:"screenshotURL"`
		} `json:"task"`
	}
	if err := json.Unmarshal(resultBody, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	verdict := "clean"
	if raw.Verdicts.Overall.Malicious {
		verdict = "malicious"
	}

	cats := raw.Verdicts.URLScan.Categories
	if cats == nil {
		cats = []string{}
	}

	meta.Success = true
	return &models.URLScanData{
		Verdict:       ptr(verdict),
		Score:         ptr(raw.Verdicts.Overall.Score),
		Malicious:     ptr(raw.Verdicts.Overall.Malicious),
		Categories:    cats,
		ScreenshotURL: raw.Task.ScreenshotURL,
	}, meta
}

File: scanner-python/app/services/urlscan.py
import httpx
import asyncio
from app.config import settings
from app.utils.normalize import utc_now
from app.models.shared import SourceMetadata

URLSCAN_SUBMIT_URL = "https://urlscan.io/api/v1/scan/"
URLSCAN_RESULT_URL = "https://urlscan.io/api/v1/result/"

async def fetch_urlscan(url: str) -> tuple[dict | None, SourceMetadata]:
    source = SourceMetadata(
        source="urlscan",
        retrieved_at=utc_now(),
        success=False
    )
    try:
        async with httpx.AsyncClient() as client:
            submit_response = await client.post(
                URLSCAN_SUBMIT_URL,
                headers={
                    "API-Key": settings.urlscan_api_key,
                    "Content-Type": "application/json"
                },
                json={"url": url, "visibility": "public"},
                timeout=10.0
            )
            submit_response.raise_for_status()
            submit_data = submit_response.json()

            scan_uuid = submit_data.get("uuid")
            if not scan_uuid:
                source.error = "No scan UUID returned"
                return None, source

            await asyncio.sleep(15)

            result_response = await client.get(
                f"{URLSCAN_RESULT_URL}{scan_uuid}/",
                timeout=10.0
            )
            result_response.raise_for_status()
            raw = result_response.json()

            verdicts = raw.get("verdicts", {}).get("overall", {})
            categories = raw.get("verdicts", {}).get("urlscan", {}).get("categories", [])
            screenshot = raw.get("task", {}).get("screenshotURL")

            normalized = {
                "verdict": "malicious" if verdicts.get("malicious") else "clean",
                "score": verdicts.get("score", 0),
                "malicious": verdicts.get("malicious", False),
                "categories": categories,
                "screenshot_url": screenshot,
            }

            source.success = True
            return normalized, source

    except httpx.TimeoutException:
        source.error = "Request timed out"
        return None, source
    except httpx.HTTPStatusError as e:
        source.error = f"HTTP {e.response.status_code}"
        return None, source
    except Exception as e:
        source.error = str(e)
        return None, source